File: chains/cpp/dexvm/test/conformance.cpp
// Copyright (C) 2026, Lux Industries, Inc. All rights reserved.
// SPDX-License-Identifier: BSD-3-Clause-Eco
//
// conformance.cpp — the C++ D-chain's answers to the shared corpus.
//
// One of the evaluators — Go, Rust and C++ — that read the same
// conformance/corpus/vectors.tsv and print the same seven fields per vector.
// The runner compares the sets of lines; this program never sees another
// implementation's answer and has nothing to agree with.
//
// D IS NOT A BLOCK CHAIN IN THIS CORPUS. Go's dexvm is a REGISTRY, and what it
// decides is identity and admission — what an asset IS, what a market IS, which
// kinds may be registered, and whether native value may activate at all. Those
// are consensus decisions with no wire of their own: two implementations
// deriving different bytes for one asset have forked the value plane without
// ever disagreeing about a transaction.
//
// So a D vector's wire column carries the ARGUMENTS, pipe-separated ASCII, and
// the op says which function is being asked:
//
//   assetid   networkID | sourceChainHex | kindToken | refHex
//   marketid  networkID | baseHex | quoteHex | venueHex
//   kind      token
//   mode      token
//   class     networkID
//   guard     valueEnabled | modeToken | capsOn | realAssetsOnly | haltReady
//
// Usage: dexvm_conformance <vectors.tsv>

#include "lux/conformance/corpus.hpp"

#include "lux/dexvm/asset.hpp"
#include "lux/dexvm/consensus_mode.hpp"
#include "lux/dexvm/gate.hpp"

#include <cstdio>
#include <string>
#include <vector>

using namespace lux::dexvm;
namespace conf = lux::conformance;

namespace {

// The class of a refusal, decided on its SENTINEL rather than on the sentence
// it was wrapped in.
//
// Go's classify unwraps to the deepest error before it reads any words, and
// prints the wrapper in the note beside it. The C++ Error carries the same
// identity as a code whose text_of() IS the Go sentinel, so reading that is
// the same question. Reading the wrapped text instead is a different one, and
// it answers differently: "canonical reference does not match asset kind: UTXO
// assetID must be 32 bytes" is a syntactic refusal whose DETAIL contains the
// word utxo, and the ledger arm claims it.
// Err::Other is this port's "no sentinel" — the refusal was raised with words
// and no identity, which is Go's fmt.Errorf without a %w. root() returns such
// an error unchanged, so its own text IS what gets classified, on both sides.
std::string classify_of(const Error& e) {
    if (e.code == Err::Other) return conf::classify(e.text);
    return conf::classify(std::string(text_of(e.code)));
}

std::vector<std::string> split(const std::string& s, char sep) {
    std::vector<std::string> out;
    std::size_t start = 0;
    for (;;) {
        const std::size_t at = s.find(sep, start);
        if (at == std::string::npos) {
            out.push_back(s.substr(start));
            return out;
        }
        out.push_back(s.substr(start, at - start));
        start = at + 1;
    }
}

bool parse_u32(const std::string& s, std::uint32_t& out) {
    if (s.empty()) return false;
    std::uint64_t v = 0;
    for (const char c : s) {
        if (c < '0' || c > '9') return false;
        v = v * 10 + static_cast<std::uint64_t>(c - '0');
        if (v > 0xFFFFFFFFull) return false;
    }
    out = static_cast<std::uint32_t>(v);
    return true;
}

bool parse_bool(const std::string& s, bool& out) {
    if (s == "true") {
        out = true;
        return true;
    }
    if (s == "false") {
        out = false;
        return true;
    }
    return false;
}

bool parse_id(const std::string& s, Id& out) {
    std::vector<std::uint8_t> b;
    if (!conf::unhex(s, b) || b.size() != out.size()) return false;
    std::copy(b.begin(), b.end(), out.begin());
    return true;
}

// A vector whose ARGUMENTS do not read. That is this evaluator's failure, not
// the chain's, and it is reported as such rather than as a refusal the chain
// never made.
conf::Row bad_input(conf::Row r, const std::string& why) {
    r.parse = conf::kInternal;
    r.syntactic = conf::kInternal;
    r.exec = conf::kInternal;
    r.note = why;
    return r;
}

conf::Row eval_assetid(conf::Row r, const std::vector<std::string>& a) {
    r.kind = "AssetID";
    if (a.size() != 4) return bad_input(r, "assetid wants 4 arguments");
    std::uint32_t network = 0;
    Id chain{};
    std::vector<std::uint8_t> ref;
    if (!parse_u32(a[0], network)) return bad_input(r, "network id: " + a[0]);
    if (!parse_id(a[1], chain)) return bad_input(r, "source chain: " + a[1]);
    if (!conf::unhex(a[3], ref)) return bad_input(r, "reference: " + a[3]);

    // The kind token goes through the registry's own parser, so a token it
    // refuses is refused at the layer a manifest would meet it.
    auto kind = parse_kind(a[2]);
    if (!kind) {
        r.syntactic = classify_of(kind.error());
        r.exec = r.syntactic;
        r.note = kind.error().text;
        return r;
    }
    r.syntactic = conf::kOk;

    auto asset = derive_asset_id(network, chain, kind.value(), ref);
    if (!asset) {
        r.exec = classify_of(asset.error());
        r.note = asset.error().text;
        return r;
    }
    const Id& v = asset.value();
    r.hash = conf::hex(v.data(), v.size());
    r.exec = conf::kOk;
    r.note = "derived";
    return r;
}

conf::Row eval_marketid(conf::Row r, const std::vector<std::string>& a) {
    r.kind = "MarketID";
    if (a.size() != 4) return bad_input(r, "marketid wants 4 arguments");
    std::uint32_t network = 0;
    Id base{}, quote{};
    std::vector<std::uint8_t> venue;
    if (!parse_u32(a[0], network)) return bad_input(r, "network id: " + a[0]);
    if (!parse_id(a[1], base)) return bad_input(r, "base asset: " + a[1]);
    if (!parse_id(a[2], quote)) return bad_input(r, "quote asset: " + a[2]);
    if (!conf::unhex(a[3], venue)) return bad_input(r, "venue: " + a[3]);

    r.syntactic = conf::kOk;
    const Id m = market_id(network, base, quote, venue);
    r.hash = conf::hex(m.data(), m.size());
    r.exec = conf::kOk;
    r.note = "derived";
    return r;
}

conf::Row eval_kind(conf::Row r, const std::vector<std::string>& a) {
    r.kind = "AssetKind";
    if (a.size() != 1) return bad_input(r, "kind wants 1 argument");
    auto k = parse_kind(a[0]);
    if (!k) {
        r.syntactic = classify_of(k.error());
        r.exec = r.syntactic;
        r.note = k.error().text;
        return r;
    }
    r.syntactic = conf::kOk;
    r.exec = conf::kOk;
    r.note = std::string(to_string(k.value()));
    return r;
}

conf::Row eval_mode(conf::Row r, const std::vector<std::string>& a) {
    r.kind = "ConsensusMode";
    if (a.size() != 1) return bad_input(r, "mode wants 1 argument");
    auto m = parse_consensus_mode(a[0]);
    if (!m) {
        r.syntactic = classify_of(m.error());
        r.exec = r.syntactic;
        r.note = m.error().text;
        return r;
    }
    r.syntactic = conf::kOk;
    r.exec = conf::kOk;
    r.note = std::string(to_string(m.value()));
    return r;
}

conf::Row eval_class(conf::Row r, const std::vector<std::string>& a) {
    r.kind = "NetworkClass";
    if (a.size() != 1) return bad_input(r, "class wants 1 argument");
    std::uint32_t network = 0;
    if (!parse_u32(a[0], network)) return bad_input(r, "network id: " + a[0]);
    r.syntactic = conf::kOk;
    r.exec = conf::kOk;
    r.note = std::string(to_string(network_class_for(network)));
    return r;
}

conf::Row eval_guard(conf::Row r, const std::vector<std::string>& a) {
    r.kind = "ValueActivation";
    if (a.size() != 5) return bad_input(r, "guard wants 5 arguments");
    bool enabled = false, caps = false, real = false, halt = false;
    if (!parse_bool(a[0], enabled)) return bad_input(r, "value enabled: " + a[0]);
    if (!parse_bool(a[2], caps)) return bad_input(r, "caps: " + a[2]);
    if (!parse_bool(a[3], real)) return bad_input(r, "real assets: " + a[3]);
    if (!parse_bool(a[4], halt)) return bad_input(r, "halt ready: " + a[4]);

    // An unparseable mode token is UNSET, because that is what a caller who
    // declared nothing legible has declared — and UNSET is refused. Reading it
    // as an error here would report the parse rather than the guard.
    ConsensusMode mode = ConsensusMode::Unset;
    if (auto m = parse_consensus_mode(a[1])) mode = m.value();
    r.syntactic = conf::kOk;

    auto st = guard_value_activation(enabled, mode, LaunchAssertions{caps, real, halt});
    if (!st) {
        r.exec = classify_of(st.error());
        r.note = st.error().text;
        return r;
    }
    r.exec = conf::kOk;
    const std::string mode_name(to_string(st.value().mode));
    r.note = st.value().status.empty() ? mode_name + " (no disclaimer)"
                                       : mode_name + " " + st.value().status;
    return r;
}

}  // namespace

int main(int argc, char** argv) {
    if (argc < 2) {
        std::fprintf(stderr, "usage: dexvm_conformance <vectors.tsv>\n");
        return 2;
    }
    std::vector<conf::Vector> vectors;
    if (!conf::read(argv[1], "dexvm_conformance", vectors)) return 1;

    for (const auto& v : vectors) {
        if (v.chain != "D") continue;
        conf::Row r;
        r.id = v.id;
        r.parse = "ok";
        // The wire column is hex, always — a D vector's arguments are ASCII,
        // but they travel the same column a block's bytes do, so they are
        // hex-encoded like everything else. Splitting the column itself finds
        // one field whose text is "4552433230".
        std::vector<std::uint8_t> raw;
        if (!conf::unhex(v.wire, raw)) {
            conf::print(bad_input(r, "corpus wire is not hex"));
            continue;
        }
        const std::vector<std::string> args =
            split(std::string(raw.begin(), raw.end()), '|');

        if (v.op == "assetid") {
            conf::print(eval_assetid(r, args));
        } else if (v.op == "marketid") {
            conf::print(eval_marketid(r, args));
        } else if (v.op == "kind") {
            conf::print(eval_kind(r, args));
        } else if (v.op == "mode") {
            conf::print(eval_mode(r, args));
        } else if (v.op == "class") {
            conf::print(eval_class(r, args));
        } else if (v.op == "guard") {
            conf::print(eval_guard(r, args));
        } else {
            std::fprintf(stderr, "dexvm_conformance: unknown op %s on %s\n", v.op.c_str(),
                         v.id.c_str());
            return 1;
        }
    }
    return 0;
}

File: chains/rust/dexvm/src/bin/conformance.rs
// SPDX-License-Identifier: BSD-3-Clause-Eco
// Copyright (C) 2019-2026, Lux Industries Inc. All rights reserved.

//! The Rust D-chain's answers to the shared corpus.
//!
//! One of the evaluators — Go, Rust, C++ — that read the same
//! `conformance/corpus/vectors.tsv` and print the same seven fields per vector.
//! The runner compares their lines; this program never sees another
//! implementation's answer and has nothing to agree with.
//!
//! Usage: `conformance <vectors.tsv>`
//!
//! It answers the `D` rows and stays silent on every other chain's, because a
//! row is a claim and this program has no claim to make about a P-chain
//! transaction.
//!
//! D IS NOT A BLOCK CHAIN HERE, so a D vector's wire column carries the
//! ARGUMENTS to a derivation rather than a serialization: pipe-separated ASCII,
//! hex-encoded because it travels the column a block's bytes travel. The op says
//! which function is being asked.
//!
//! ```text
//! assetid   networkID | sourceChainHex | kindToken | refHex
//! marketid  networkID | baseHex | quoteHex | venueHex
//! kind      token
//! mode      token
//! class     networkID
//! guard     valueEnabled | modeToken | capsOn | realAssetsOnly | haltReady
//! ```

use lux_dexvm::asset::{derive_asset_id, market_id, AssetKind};
use lux_dexvm::error::Error;
use lux_dexvm::ids::{self, Id};
use lux_dexvm::mode::{guard_value_activation, ConsensusMode, LaunchAssertions};
use lux_dexvm::network::network_class_for;

const NONE: &str = "-";

// The verdict vocabulary, shared with the Go and C++ evaluators. A word here
// means the same thing in all three, which is what lets the runner compare
// strings instead of understanding rules.
const OK: &str = "OK";
const SYNTACTIC: &str = "SYNTACTIC";
const OVERFLOW: &str = "OVERFLOW";
const LEDGER: &str = "LEDGER";
const AUTH: &str = "AUTH";
const WARP: &str = "WARP";
const UNSUPPORTED: &str = "UNSUPPORTED";
const INTERNAL: &str = "INTERNAL";

fn main() {
    let path = std::env::args().nth(1).unwrap_or_else(|| {
        eprintln!("usage: conformance <vectors.tsv>");
        std::process::exit(2);
    });
    let text = match std::fs::read_to_string(&path) {
        Ok(t) => t,
        Err(e) => {
            eprintln!("conformance: {path}: {e}");
            std::process::exit(1);
        }
    };

    for line in text.lines() {
        if line.is_empty() || line.starts_with('#') {
            continue;
        }
        let f: Vec<&str> = line.split('\t').collect();
        if f.len() != 5 || f[0] != "V" {
            eprintln!("conformance: not a vector line: {line}");
            std::process::exit(1);
        }
        if f[2] != "D" {
            continue;
        }
        println!("{}", evaluate(f[1], f[3], f[4]));
    }
}

struct Row {
    id: String,
    parse: String,
    kind: String,
    hash: String,
    syntactic: String,
    exec: String,
    note: String,
}

impl std::fmt::Display for Row {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut note = self.note.replace(['\t', '\n'], " ");
        note.truncate(floor_char_boundary(&note, 160));
        let note = if note.is_empty() { NONE } else { &note };
        write!(
            f,
            "R\t{}\t{}\t{}\t{}\t{}\t{}\t{}",
            self.id, self.parse, self.kind, self.hash, self.syntactic, self.exec, note
        )
    }
}

/// The largest index at or below `at` that a `String` may be cut at. The
/// reference trims a note to 160 BYTES; cutting a multi-byte character in half
/// would panic here, and the refusal that carries an em dash is one of the
/// notes this trims.
fn floor_char_boundary(s: &str, at: usize) -> usize {
    if at >= s.len() {
        return s.len();
    }
    let mut i = at;
    while !s.is_char_boundary(i) {
        i -= 1;
    }
    i
}

fn row(id: &str) -> Row {
    Row {
        id: id.to_string(),
        parse: NONE.into(),
        kind: NONE.into(),
        hash: NONE.into(),
        syntactic: NONE.into(),
        exec: NONE.into(),
        note: String::new(),
    }
}

fn evaluate(id: &str, op: &str, wire: &str) -> Row {
    let mut r = row(id);
    // The arguments are ASCII inside the hex, so the column is decoded once and
    // then split. Splitting the column itself would find one field whose text is
    // "4552433230".
    let Some(raw) = ids::unhex(wire_or_empty(wire)) else {
        return bad_input(r, "corpus wire is not hex");
    };
    let Ok(text) = String::from_utf8(raw) else {
        return bad_input(r, "corpus wire is not an argument list");
    };
    let args: Vec<&str> = text.split('|').collect();
    r.parse = "ok".into();

    match op {
        "assetid" => eval_assetid(r, &args),
        "marketid" => eval_marketid(r, &args),
        "kind" => eval_kind(r, &args),
        "mode" => eval_mode(r, &args),
        "class" => eval_class(r, &args),
        "guard" => eval_guard(r, &args),
        _ => {
            r.parse = INTERNAL.into();
            r.syntactic = INTERNAL.into();
            r.exec = INTERNAL.into();
            r.note = format!("unknown op {op}");
            r
        }
    }
}

/// A vector with no bytes writes `-`, which is an empty argument list rather
/// than an unreadable one.
fn wire_or_empty(wire: &str) -> &str {
    if wire == NONE {
        ""
    } else {
        wire
    }
}

/// A vector whose ARGUMENTS do not read. That is this evaluator's failure, not
/// the chain's, and it is reported as such rather than as a refusal the chain
/// never made.
fn bad_input(mut r: Row, why: &str) -> Row {
    r.parse = INTERNAL.into();
    r.syntactic = INTERNAL.into();
    r.exec = INTERNAL.into();
    r.note = why.into();
    r
}

fn eval_assetid(mut r: Row, a: &[&str]) -> Row {
    r.kind = "AssetID".into();
    if a.len() != 4 {
        return bad_input(r, &format!("assetid wants 4 arguments, got {}", a.len()));
    }
    let Some(network) = u32_of(a[0]) else {
        return bad_input(r, &format!("network id: {}", a[0]));
    };
    let Some(chain) = id_of(a[1]) else {
        return bad_input(r, &format!("source chain: {}", a[1]));
    };
    let Some(reference) = ids::unhex(a[3]) else {
        return bad_input(r, &format!("reference: {}", a[3]));
    };

    // The kind token goes through the registry's own parser, so a token it
    // refuses is refused at the layer a manifest would meet it.
    let kind = match a[2].parse::<AssetKind>() {
        Ok(k) => k,
        Err(e) => return refused_at_syntax(r, &e),
    };
    r.syntactic = OK.into();

    match derive_asset_id(network, chain, kind, &reference) {
        Ok(asset) => {
            r.hash = ids::hex(&asset);
            r.exec = OK.into();
            r.note = "derived".into();
        }
        Err(e) => {
            r.exec = classify(&e.sentinel()).into();
            r.note = e.to_string();
        }
    }
    r
}

fn eval_marketid(mut r: Row, a: &[&str]) -> Row {
    r.kind = "MarketID".into();
    if a.len() != 4 {
        return bad_input(r, &format!("marketid wants 4 arguments, got {}", a.len()));
    }
    let Some(network) = u32_of(a[0]) else {
        return bad_input(r, &format!("network id: {}", a[0]));
    };
    let Some(base) = id_of(a[1]) else {
        return bad_input(r, &format!("base asset: {}", a[1]));
    };
    let Some(quote) = id_of(a[2]) else {
        return bad_input(r, &format!("quote asset: {}", a[2]));
    };
    let Some(venue) = ids::unhex(a[3]) else {
        return bad_input(r, &format!("venue: {}", a[3]));
    };

    r.syntactic = OK.into();
    r.hash = ids::hex(&market_id(network, base, quote, &venue));
    r.exec = OK.into();
    r.note = "derived".into();
    r
}

fn eval_kind(mut r: Row, a: &[&str]) -> Row {
    r.kind = "AssetKind".into();
    if a.len() != 1 {
        return bad_input(r, &format!("kind wants 1 argument, got {}", a.len()));
    }
    match a[0].parse::<AssetKind>() {
        Ok(k) => {
            r.syntactic = OK.into();
            r.exec = OK.into();
            r.note = k.to_string();
            r
        }
        Err(e) => refused_at_syntax(r, &e),
    }
}

fn eval_mode(mut r: Row, a: &[&str]) -> Row {
    r.kind = "ConsensusMode".into();
    if a.len() != 1 {
        return bad_input(r, &format!("mode wants 1 argument, got {}", a.len()));
    }
    match a[0].parse::<ConsensusMode>() {
        Ok(m) => {
            r.syntactic = OK.into();
            r.exec = OK.into();
            r.note = m.to_string();
            r
        }
        Err(e) => refused_at_syntax(r, &e),
    }
}

fn eval_class(mut r: Row, a: &[&str]) -> Row {
    r.kind = "NetworkClass".into();
    if a.len() != 1 {
        return bad_input(r, &format!("class wants 1 argument, got {}", a.len()));
    }
    let Some(network) = u32_of(a[0]) else {
        return bad_input(r, &format!("network id: {}", a[0]));
    };
    r.syntactic = OK.into();
    r.exec = OK.into();
    r.note = network_class_for(network).to_string();
    r
}

fn eval_guard(mut r: Row, a: &[&str]) -> Row {
    r.kind = "ValueActivation".into();
    if a.len() != 5 {
        return bad_input(r, &format!("guard wants 5 arguments, got {}", a.len()));
    }
    let Some(enabled) = bool_of(a[0]) else {
        return bad_input(r, &format!("value enabled: {}", a[0]));
    };
    let Some(caps_on) = bool_of(a[2]) else {
        return bad_input(r, &format!("caps: {}", a[2]));
    };
    let Some(real_assets_only) = bool_of(a[3]) else {
        return bad_input(r, &format!("real assets: {}", a[3]));
    };
    let Some(halt_ready) = bool_of(a[4]) else {
        return bad_input(r, &format!("halt ready: {}", a[4]));
    };

    // An unparseable mode token is UNSET, because that is what a caller who
    // declared nothing legible has declared — and UNSET is refused. Reading it
    // as an error here would report the parse rather than the guard.
    let mode = a[1].parse::<ConsensusMode>().unwrap_or_default();
    r.syntactic = OK.into();

    match guard_value_activation(
        enabled,
        mode,
        LaunchAssertions {
            caps_on,
            real_assets_only,
            halt_ready,
        },
    ) {
        Ok(status) => {
            r.exec = OK.into();
            r.note = if status.status.is_empty() {
                format!("{} (no disclaimer)", status.mode)
            } else {
                format!("{} {}", status.mode, status.status)
            };
        }
        Err(e) => {
            r.exec = classify(&e.sentinel()).into();
            r.note = e.to_string();
        }
    }
    r
}

/// A refusal the chain reached on the arguments' own shape, before it derived
/// anything. Both layers carry it, exactly as the reference does: a token that
/// is not a kind is refused at the layer a manifest would meet it, and there is
/// nothing left for the derivation to answer.
fn refused_at_syntax(mut r: Row, e: &Error) -> Row {
    r.syntactic = classify(&e.sentinel()).into();
    r.exec = r.syntactic.clone();
    r.note = e.to_string();
    r
}

// Reading one argument. These are the HARNESS's readers, not the chain's: what
// they refuse becomes INTERNAL, a loud non-answer the runner reports, never a
// verdict. So they accept exactly what the generator writes — decimal digits,
// `true`/`false`, an even run of hex — and nothing else. Go's readers are
// looser here (`ParseBool` also takes `1` and `t`, `ParseUint` also takes a
// leading `+`), and the looser direction is the wrong one for a reader: it
// would answer a verdict for an argument list nobody wrote.

fn u32_of(s: &str) -> Option<u32> {
    if s.is_empty() || !s.bytes().all(|b| b.is_ascii_digit()) {
        return None;
    }
    s.parse::<u32>().ok()
}

fn id_of(s: &str) -> Option<Id> {
    ids::from_slice(&ids::unhex(s)?)
}

fn bool_of(s: &str) -> Option<bool> {
    match s {
        "true" => Some(true),
        "false" => Some(false),
        _ => None,
    }
}

/// The same word table the Go and C++ evaluators carry, in the same order, so
/// that one refusal means one class in all three.
///
/// It is handed the refusal's SENTINEL, never its sentence. The two differ on
/// purpose here: `canonical reference does not match asset kind: UTXO assetID
/// must be 32 bytes` is a refusal about a reference's shape whose detail
/// contains the word `utxo`, and the ledger arm would claim it. Go reads the
/// same half — `classify` unwraps to the root error before it reads any words.
///
/// The order is load-bearing for the same kind of reason: a refusal that names a
/// kind the chain does not run usually lists the ones it does, and those lists
/// carry ledger words.
fn classify(s: &str) -> &'static str {
    let s = s.to_lowercase();
    let has = |w: &str| s.contains(w);
    if has("overflow") || has("underflow") {
        OVERFLOW
    } else if has("wrong transaction type")
        || has("wrong tx type")
        || has("not permitted")
        || has("not held")
        || has("unsupported")
        || has("forbidden")
        || has("unknown")
        || has("parameter set")
    {
        UNSUPPORTED
    } else if has("credential")
        || has("signature")
        || has("unauthorized")
        || has("not authorised")
        || has("not authorized")
        || has("proof verification failed")
        || has("does not match auth")
    {
        AUTH
    } else if has("warp") {
        WARP
    } else if has("utxo")
        || has("funds")
        || has("insufficient")
        || has("burn")
        || has("consumed")
        || has("produced")
        || has("flow")
        || has("fee")
        || has("not found")
        || has("doesn't exist")
        || has("does not exist")
        || has("isn't a current")
        || has("not validator")
        || has("no such")
        || has("could not load")
        || has("shared memory")
        || has("state root")
        || has("precedes its parent")
        || has("skew allowance")
    {
        LEDGER
    } else {
        SYNTACTIC
    }
}
